File: 3242/main.go
package main

import (
	"fmt"
)

func main() {
	ns1 := NeighborSum{[][]int{
		{3,6,0},
		{8,5,1},
		{2,4,7},
	}}
	fmt.Println(ns1.AdjacentSum(1))
	fmt.Println(ns1.AdjacentSum(3))
	fmt.Println(ns1.AdjacentSum(7))
	fmt.Println(ns1.AdjacentSum(2))
	fmt.Println(ns1.AdjacentSum(5))
	fmt.Println(ns1.AdjacentSum(8))
	fmt.Println(ns1.AdjacentSum(0))
	fmt.Println(ns1.DiagonalSum(1))
	fmt.Println(ns1.DiagonalSum(4))
	fmt.Println(ns1.DiagonalSum(8))
	fmt.Println(ns1.DiagonalSum(6))
	fmt.Println(ns1.DiagonalSum(3))
	fmt.Println(ns1.DiagonalSum(2))
	fmt.Println(ns1.DiagonalSum(6))
	fmt.Println(ns1.DiagonalSum(5))
	fmt.Println(ns1.DiagonalSum(7))
	fmt.Println(ns1.DiagonalSum(4))
	fmt.Println(ns1.DiagonalSum(0))

	ns2 := NeighborSum{[][]int{
		{1,2,0,3},
		{4,7,15, 6},
		{8,9,10,11},
		{12,13,14,5},
	}}
	fmt.Println(ns2.AdjacentSum(15))
	fmt.Println(ns2.DiagonalSum(9))

	ns := NeighborSum{[][]int{
		{0,1,2},
		{3,4,5},
		{6,7,8},
	}}
	fmt.Println(ns.AdjacentSum(1))
	fmt.Println(ns.AdjacentSum(4))
	fmt.Println(ns.DiagonalSum(4))
	fmt.Println(ns.DiagonalSum(8))
}

type NeighborSum struct {
	grid [][]int
}

func Constructor(grid [][]int) NeighborSum {
	return NeighborSum{grid}
}

func (ns *NeighborSum) AdjacentSum(value int) int {
	sum := 0
	for i := range ns.grid {
		for j := range ns.grid[i] {
			if ns.grid[i][j] == value {
				// 上
				if i > 0 {
					sum += ns.grid[i-1][j]
				}
				// 下
				if i < len(ns.grid)-1 {
					sum += ns.grid[i+1][j]
				}
				// 左
				if j > 0 {
					sum += ns.grid[i][j-1]
				}
				// 右
				if j < len(ns.grid[i])-1 {
					sum += ns.grid[i][j+1]
				}
			}
		}
	}
	return sum
}

func (ns *NeighborSum) DiagonalSum(value int) int {
	sum := 0
	for i := range ns.grid {
		for j := range ns.grid[i] {
			if ns.grid[i][j] == value {
				// 左上
				if i > 0 && j > 0 {
					sum += ns.grid[i-1][j-1]
				}
				// 右下
				if i < len(ns.grid)-1 && j < len(ns.grid[i])-1 {
					sum += ns.grid[i+1][j+1]
				}
				// 左下
				if j > 0 && i < len(ns.grid)-1 {
					sum += ns.grid[i+1][j-1]
				}
				// 右上
				if j < len(ns.grid[i])-1 && i > 0 {
					sum += ns.grid[i-1][j+1]
				}
			}
		}
	}
	return sum
}

File: 3242/main.rs
use std::collections::HashMap;


const DIRS: [ &[(i32, i32)]; 2] = [
    &[(-1, 0), (1, 0), (0, -1), (0,1)],
    &[(-1, -1), (-1, 1), (1, -1), (1, 1)],
];

struct NeighborSum {
    grid: Vec<Vec<i32>>,
    pos: HashMap<i32, (usize, usize)>,
}

impl NeighborSum {
    fn new(grid: Vec<Vec<i32>>) -> Self {
        let mut pos = HashMap::new();
        for (i, row) in grid.iter().enumerate() {
            for (j, &val) in row.iter().enumerate() {
                pos.insert(val, (i, j));
            }
        }
        NeighborSum{grid, pos}
    }

    fn adjacent_sum(&self, value: i32) -> i32 {
        self.get_sum(value, 0)
    }

    fn diagonal_sum(&self, value: i32) -> i32 {
        self.get_sum(value, 1)
    }

    fn get_sum(&self, value: i32, idx: usize) -> i32 {
        if let Some(&(x, y)) = self.pos.get(&value) {
            let mut sum = 0;
            for &(dx, dy) in DIRS[idx] {
                let nx = x as i32 + dx;
                let ny = y as i32 + dy;
                if nx >= 0 && (nx as usize) < self.grid.len() && ny >= 0 && (ny as usize) < self.grid[0].len() {
                    sum += self.grid[nx as usize][ny as usize];
                }
            }
            sum
        } else {
            0
        }
    }
}

fn main() {
    let grid = vec![
        vec![0,1,2],
        vec![3,4,5],
        vec![6,7,8],
    ];
    let ns = NeighborSum::new(grid);
    println!("{}", ns.adjacent_sum(1));
    println!("{}", ns.adjacent_sum(4));
    println!("{}", ns.diagonal_sum(4));
    println!("{}", ns.diagonal_sum(8));
}